### PR TITLE
Fix for numpy arrays when the size 1 arrays were treated as scalars

### LINKED ISSUE #14706: Symbol violates the commutativity of addition when it sometimes converts one-element arrays to scalar
```
from sympy import symbols
import numpy as np

for i in [2, 1]:
    z = np.zeros(i)
    xz = symbols('x') + z
    print(z.shape)
    print(xz.shape)
    print()
```
prints:
```
(2,)
(2,)

(1,)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-165-14d0a2118753> in <module>()
      6     xz = symbols('x') + z
      7     print(z.shape)
----> 8     print(xz.shape)

AttributeError: 'Symbol' object has no attribute 'shape'
```

A scalar is  *definitely* not a numpy array with shape `(1,)`, `(1, 1)`, etc.
It's debatable whether a scalar is a numpy array with shape `()`, but I would simply return a numpy array in this case too.  Every library that has ever made the mistake of doing this dimension contraction had to eventually undo it (e.g., Theano).

The problem is that it's very easy to get the lone element from an array of size 1:
```
In [1]: o = np.ones([1 for _ in range(10)])

In [2]: o
Out[2]: array([[[[[[[[[[1.]]]]]]]]]])

In [3]: o.item()
Out[3]: 1.0
```
But it's very hard to go from an element and try to figure out what the shape it should have had and then give it the right shape.

Since this breaks my code, I can submit a patch within about a day if it will be accepted and if someone can point me to the right place in the code (I'm guessing tensor.py).

### PR DESCRIPTION
**Pull request by mijo2**
#### References to other Issues or PRs
Revives and closes #14720 
Fixes #14706

#### Brief description of what is fixed or changed

--> Added a bug fix in sympify whereby the numpy arrays of size 1 won't be treated as scalars

#### Other comments

Example:
```
>>> from sympy import *
>>> from sympy.abc import *
>>> import numpy
>>> z = np.zeros((1,1))
```

Before the fix:
```
>>> f = x + z
x
```

After the fix:
```
>>> f = x + z
array([[x]], dtype=object)
```

#### Release Notes

* core
  *  sympify no longer converts numpy arrays with size to scalars

### CODE REVIEW

**Comment by asmeurer** [comment]
We all the comments from the original PR addressed here?

**Comment by mijo2** [comment]
@asmeurer I think so. The main thing was changing the last test to a Sympify error check. But I will add more test cases for Sympify if required.

**Comment by czgdp1807** [comment]
More tests are needed, diff coverage is too low.

File: sympy/core/tests/test_sympify.py [hunk 1]
```diff
@@ -695,3 +696,39 @@ def test_issue_16759():
 def test_issue_17811():
     a = Function('a')
     assert sympify('a(x)*5', evaluate=False) == Mul(a(x), 5, evaluate=False)
+
+
+def test_issue_14706():
+    if not numpy:
+        skip("numpy not installed.")
+
+    from sympy import symbols
```
Comment (smichr) [inline sympy/core/tests/test_sympify.py]:
```suggestion
```

File: sympy/core/tests/test_sympify.py [hunk 2]
```diff
@@ -695,3 +696,39 @@ def test_issue_16759():
...
+
+    x = symbols('x')
```
Comment (smichr) [inline sympy/core/tests/test_sympify.py]:
```suggestion
    x = Symbol('x')
```

**Comment by mijo2** [comment]
@asmeurer @smichr I think this is ready to be merged now.

**Comment by smichr** [comment]
A comment on the orginal issue was

> It looks like the numpy people realized their mistake (numpy/numpy#10404) and are trying to deprecate it by the way, so in the long run this may be fixed on their end, and a link to the numpy issue might be good to add in a comment in the source code.

If the issue is fixed then can these changes be dropped or will these changes still be necessary. If the former then what we need is a test which passes now but will fail when numpy fixes the issue. A string in the function would indicate that this PR should be reverted at that time, e.g.
```python
def test_numpy_issue_10404():
    # if this fails then the PR in which this test was added
    # can be reverted (or whatever the instructions are)
    assert something that passes
```

**Comment by mijo2** [comment]
Ok @smichr I will take a look and add these kind of tests along with comment about what to do and which commit to PR.

**Comment by mijo2** [comment]
@smichr I have added the kind of tests along with appropriate comments for someone in future to take up the issue  if ever that happens.

**Comment by smichr** [comment]
OK, thanks for this fix.

**Pull request merged**